File: internal/model/normalize.go
package model

import (
	"math"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/ProfessorBagholder/Bagholder/internal/py"
	"github.com/ProfessorBagholder/Bagholder/internal/store"
	"github.com/ProfessorBagholder/Bagholder/internal/symbols"
)

func compact(s string) string { return symbols.Compact(s) }

func normAccountName(s string) string { return symbols.NormAccountName(s) }

func optionRight(symbol string) string { return symbols.Right(symbol) }

func optionExpiry(symbol string) string { return symbols.Expiry(symbol) }

func isMultileg(a *Act) bool { return strings.Contains(compact(a.RawType), "MULTILEG") }

type rollKey struct{ account, underlying, right string }

func rollKeyOf(a *Act) rollKey {
	return rollKey{fifoAccount(a), underlying(a.Symbol), optionRight(a.Symbol)}
}

func daysBetween(a, b string) int {
	da, oka := py.ParseDate(a)
	db, okb := py.ParseDate(b)
	if !oka || !okb {
		return 0
	}
	d := int(db.Sub(da).Hours() / 24)
	if d < 0 {
		return 0
	}
	return d
}

func shiftDate(iso string, days int) string {
	t, ok := py.ParseDate(iso)
	if !ok {
		if len(iso) > 10 {
			return iso[:10]
		}
		return iso
	}
	return py.DateStr(t.AddDate(0, 0, days))
}

func whenParts(occurred string) (string, string) {
	s := py.Strip(occurred)
	if s == "" {
		return "", ""
	}
	if !strings.Contains(s, "T") {
		return cut(s, 10), ""
	}
	if strings.HasSuffix(s, "Z") {
		s = s[:len(s)-1] + "+00:00"
	}
	t, naive, ok := py.ParseISO(s)
	if !ok {
		return cut(s, 10), ""
	}
	if naive {
		t = time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), time.UTC)
	}
	loc := t.In(TimeTZ)
	return loc.Format("2006-01-02"), loc.Format("15:04")
}

func cut(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

func fmt8(v float64) string { return py.Fixed(v, 8) }

func isCryptoActivity(a *Act) bool {
	return strings.HasPrefix(compact(a.RawType), "CRYPTO") || strings.HasPrefix(compact(a.ActivityType), "CRYPTO")
}

func kindOf(a *Act) string {
	if py.Contains(Kinds, a.Kind) {
		return a.Kind
	}
	if isCryptoActivity(a) {
		return "Crypto"
	}
	if isOption(a.Symbol) {
		return "Options"
	}
	return "Shares"
}

func isIntentionalOpen(a *Act) bool {
	for _, f := range []string{compact(a.ActivityType), compact(a.ActivitySubType)} {
		if strings.Contains(f, "TOOPEN") {
			return true
		}
	}
	for _, f := range []string{compact(a.ActivityType), compact(a.ActivitySubType)} {
		if f == "STO" || f == "BTO" {
			return true
		}
	}
	return false
}

func isCloseOnly(a *Act) bool {
	fields := []string{compact(a.ActivityType), compact(a.ActivitySubType)}
	for _, f := range fields {
		if strings.Contains(f, "TOCLOSE") || f == "BTC" || f == "STC" {
			return true
		}
	}
	for _, f := range fields {
		if strings.Contains(f, "EXPIR") || strings.Contains(f, "ASSIGN") || strings.Contains(f, "EXERCISE") {
			return true
		}
	}
	return false
}

func openingDirection(a *Act, side string) string {
	if side == "BUY" {
		if isCloseOnly(a) {
			return ""
		}
		return "LONG"
	}
	if side == "SELL" {
		if isOption(a.Symbol) {
			if isCloseOnly(a) {
				return ""
			}
			return "SHORT"
		}
		if isIntentionalOpen(a) {
			return "SHORT"
		}
		return ""
	}
	return ""
}

func NormalizeActivity(activity *Act) *Act {
	c := activity.Clone()
	a := &c
	a.Normalized = true
	a.AccountType = normAccountName(a.AccountType)
	rt := compact(a.RawType)
	at := compact(a.ActivityType)
	cash := a.NetCashAmount
	qty := math.Abs(a.Quantity)
	a.Flags = []string{}

	if rt == "CRYPTOBUY" || at == "CRYPTOBUY" {
		a.Category, a.ActivityType, a.ActivitySubType, a.Kind = "trade", "Trade", "BUY", "Crypto"
		a.Quantity = qty
		a.NetCashAmount = -math.Abs(cash)
		return a
	}
	if rt == "CRYPTOSELL" || at == "CRYPTOSELL" {
		a.Category, a.ActivityType, a.ActivitySubType, a.Kind = "trade", "Trade", "SELL", "Crypto"
		a.Quantity = -qty
		a.NetCashAmount = math.Abs(cash)
		return a
	}
	if rt == "CRYPTOTRANSFER" || at == "CRYPTOTRANSFER" {
		sub := compact(a.ActivitySubType)
		a.Category, a.ActivityType, a.Kind = "trade", "Trade", "Crypto"
		a.Flags = append(a.Flags, "transfer")
		if strings.Contains(sub, "OUT") || cash < 0 {
			a.ActivitySubType = "SELL"
			a.Flags = append(a.Flags, "transfer-out")
			a.Quantity = -qty
			a.NetCashAmount = math.Abs(cash)
		} else {
			a.ActivitySubType = "BUY"
			a.Quantity = qty
			a.NetCashAmount = -math.Abs(cash)
		}
		return a
	}
	if rt == "CRYPTOSTAKINGREWARD" || at == "CRYPTOSTAKINGREWARD" {
		a.Category, a.ActivityType, a.ActivitySubType, a.Kind = "trade", "Trade", "BUY", "Crypto"
		a.Flags = append(a.Flags, "reward")
		a.Quantity = qty
		a.UnitPrice = 0
		a.NetCashAmount = 0
		return a
	}
	if strings.HasPrefix(rt, "CRYPTO") {
		a.Category = "other"
		a.Kind = "Crypto"
		return a
	}

	if at == "STKDIS" && rt == "DIVIDEND" && math.Abs(cash) < EPS {
		a.Category = "other"
		a.Flags = append(a.Flags, "pending-distribution")
		return a
	}

	raw := rt + at
	if strings.Contains(raw, "MULTILEG") {
		a.Category = "trade"
		if cash < 0 || compact(a.Direction) == "DEBIT" {
			a.ActivityType = "OPTIONS_BUY"
			a.ActivitySubType = "BUYTOCLOSE"
		} else {
			a.ActivityType = "OPTIONS_SELL"
			a.ActivitySubType = "SELLTOOPEN"
		}
	} else if strings.Contains(raw, "EXPIR") || strings.Contains(raw, "ASSIGN") || strings.Contains(raw, "EXERCISE") {
		a.Category = "option_event"
		if strings.Contains(raw, "ASSIGN") {
			a.ActivityType = "ASSIGN"
			a.ActivitySubType = "BUYTOCLOSE"
			a.UnitPrice = 0
		} else if strings.Contains(raw, "SHORTEXPIR") {
			a.ActivityType = "EXPIR"
			a.ActivitySubType = "BUY"
		} else if strings.Contains(raw, "EXPIR") {
			a.ActivityType = "EXPIR"
			a.ActivitySubType = "SELL"
		} else {
			a.ActivityType = "EXERCISE"
			a.ActivitySubType = "SELL"
		}
		if strings.Contains(raw, "ASSIGN") || math.Abs(cash) < 1e-12 {
			a.UnitPrice = 0
		}
		if qty > 0 {
			if a.ActivitySubType == "SELL" {
				a.Quantity = -qty
			} else {
				a.Quantity = qty
			}
		}
	}
	a.Kind = kindOf(a)
	return a
}

func NormalizeActivities(activities []store.Activity) []*Act {
	out := make([]*Act, 0, len(activities))
	for i := range activities {
		out = append(out, NormalizeActivity(&activities[i]))
	}
	return out
}

type stkdisKey struct{ symbol, date, currency string }

func foldStkdis(activities []*Act) []*Act {
	rest := make([]*Act, 0, len(activities))
	type group struct {
		pos, neg float64
		sample   *Act
	}
	groups := map[stkdisKey]*group{}
	var order []stkdisKey
	for _, a := range activities {
		if compact(a.ActivityType) != "STKDIS" {
			rest = append(rest, a)
			continue
		}
		k := stkdisKey{a.Symbol, a.TransactionDate, a.Currency}
		g, ok := groups[k]
		if !ok {
			g = &group{sample: a}
			groups[k] = g
			order = append(order, k)
		}
		q := a.Quantity
		if a.ActivitySubType == "SELL" || q < 0 {
			g.neg += math.Abs(q)
		} else {
			g.pos += math.Abs(q)
		}
	}
	for _, k := range order {
		g := groups[k]
		net := g.pos - g.neg
		if net > EPS {
			c := g.sample.Clone()
			c.Quantity = net
			c.ActivitySubType = "BUY"
			c.UnitPrice = 0
			c.NetCashAmount = 0
			c.Category = "trade"
			rest = append(rest, &c)
		}
	}
	return rest
}

type acctSym struct{ account, symbol string }

type splitKey struct{ account, symbol, day string }

func splitMarkers(activities []*Act) map[splitKey]float64 {
	out := map[splitKey]float64{}
	byBook := map[acctSym][]*Act{}
	for _, a := range activities {
		if (a.Category != "trade" && a.Category != "option_event") || a.Symbol == "" {
			continue
		}
		k := acctSym{fifoAccount(a), a.Symbol}
		byBook[k] = append(byBook[k], a)
	}
	for _, a := range activities {
		if compact(a.ActivityType) != "STKDIS" || compact(a.RawType) != "CORPORATEACTION" {
			continue
		}
		if math.Abs(a.Quantity) > EPS {
			continue
		}
		day := a.TransactionDate
		key := acctSym{fifoAccount(a), a.Symbol}
		var priced []*Act
		for _, x := range byBook[key] {
			if x.UnitPrice > 0 && compact(x.ActivityType) != "STKDIS" {
				priced = append(priced, x)
			}
		}
		sort.SliceStable(priced, func(i, j int) bool {
			if priced[i].TransactionDate != priced[j].TransactionDate {
				return priced[i].TransactionDate < priced[j].TransactionDate
			}
			return priced[i].OccurredAt < priced[j].OccurredAt
		})
		var before, after []float64
		for _, x := range priced {
			if x.TransactionDate < day {
				before = append(before, x.UnitPrice)
			} else {
				after = append(after, x.UnitPrice)
			}
		}
		if len(before) > 3 {
			before = before[len(before)-3:]
		}
		if len(after) > 3 {
			after = after[:3]
		}
		if len(before) == 0 || len(after) == 0 {
			continue
		}
		sort.Float64s(before)
		sort.Float64s(after)
		pre := before[len(before)/2]
		post := after[len(after)/2]
		if !(pre > 0) || !(post > 0) {
			continue
		}
		ratio := post / pre
		var n int
		var factor float64
		if ratio >= 1.5 {
			n = py.RoundInt(ratio)
			factor = 1.0 / float64(n)
		} else if ratio <= 1/1.5 {
			n = py.RoundInt(1 / ratio)
			factor = float64(n)
		} else {
			continue
		}
		if n < 2 || math.Abs(ratio-(1/factor))/(1/factor) > 0.35 {
			continue
		}
		out[splitKey{fifoAccount(a), a.Symbol, day}] = factor
	}
	return out
}

func fifoAccount(a *Act) string {
	nick := normAccountName(a.AccountType)
	if nick != "" {
		return nick
	}
	if a.FifoID != "" {
		return a.FifoID
	}
	return a.AccountID
}

func bookKey(a *Act) string {
	return fifoAccount(a) + "::" + a.Symbol + "::" + a.Currency
}

var removalRE = regexp.MustCompile(`CODECHANGE|SYMBOLCHANGE|TICKERCHANGE|LISTINGSTATUS|SECURITYSWAP`)

type bookID struct{ account, symbol, currency string }

type replacementIndex struct {
	removed map[bookID]string
	trades  map[bookID][]string
}

func buildReplacementIndex(activities []*Act) *replacementIndex {
	idx := &replacementIndex{removed: map[bookID]string{}, trades: map[bookID][]string{}}
	for _, a := range activities {
		key := bookID{fifoAccount(a), a.Symbol, a.Currency}
		t := compact(a.ActivityType)
		d := a.TransactionDate
		if t == "STKDIS" {
			sub := compact(a.ActivitySubType)
			if sub == "SELL" || a.Quantity < 0 {
				if cur, ok := idx.removed[key]; d != "" && (!ok || d < cur) {
					idx.removed[key] = d
				}
			}
			continue
		}
		raw := compact(a.RawType) + compact(a.AftType)
		if removalRE.MatchString(raw) {
			if cur, ok := idx.removed[key]; d != "" && (!ok || d < cur) {
				idx.removed[key] = d
			}
		}
		if (a.Category == "trade" || a.Category == "option_event") && store.TradeSide(a) != "" {
			idx.trades[key] = append(idx.trades[key], d)
		}
	}
	return idx
}

func (idx *replacementIndex) tickerWasReplaced(account, symbol, currency, byDate string) bool {
	key := bookID{account, symbol, currency}
	removedOn, ok := idx.removed[key]
	if !ok || removedOn == "" || removedOn > byDate {
		return false
	}
	for _, d := range idx.trades[key] {
		if d > removedOn {
			return false
		}
	}
	return true
}

type fill struct {
	a             *Act
	side          string
	qty           float64
	rollDirection string
	rtBefore      string
}

func setFillSide(f *fill, side, sub string) {
	f.side = side
	f.a.ActivitySubType = sub
	q := math.Abs(f.a.Quantity)
	if q > 0 {
		if side == "SELL" {
			f.a.Quantity = -q
		} else {
			f.a.Quantity = q
		}
	}
}

type dirs struct{ long, short float64 }

func (d *dirs) at(direction string) *float64 {
	if direction == "SHORT" {
		return &d.short
	}
	return &d.long
}

func resolveOptionFillSide(f *fill, rem *dirs) {
	a := f.a
	raw := compact(a.RawType) + compact(a.ActivityType)
	expirish := strings.Contains(raw, "EXPIR") || strings.Contains(raw, "ASSIGN") || strings.Contains(raw, "EXERCISE")
	if expirish {
		if strings.Contains(raw, "ASSIGN") || strings.Contains(raw, "SHORTEXPIR") {
			sub := "BUY"
			if strings.Contains(raw, "ASSIGN") {
				sub = "BUYTOCLOSE"
			}
			setFillSide(f, "BUY", sub)
		} else if strings.Contains(raw, "EXPIR") && !strings.Contains(raw, "SHORT") {
			setFillSide(f, "SELL", "SELL")
		} else if f.side == "BUY" && rem.long > EPS && rem.short <= EPS {
			setFillSide(f, "SELL", "SELL")
		} else if f.side == "SELL" && rem.short > EPS && rem.long <= EPS {
			setFillSide(f, "BUY", "BUY")
		}
		return
	}
	if !(strings.Contains(raw, "MULTILEG") || isCloseOnly(a)) {
		return
	}
	if f.side == "BUY" {
		if rem.short > EPS {
			a.ActivitySubType = "BUYTOCLOSE"
		} else {
			a.ActivitySubType = "BUYTOOPEN"
		}
	} else if f.side == "SELL" {
		if rem.long > EPS {
			a.ActivitySubType = "SELLTOCLOSE"
		} else {
			a.ActivitySubType = "SELLTOOPEN"
		}
	}
}

func isCleanOptionQty(cash, qty float64) bool {
	if !(qty > 0) {
		return false
	}
	px := math.Abs(cash) / (qty * 100.0)
	if px < 0 {
		return false
	}
	if math.Abs(px*100-math.RoundToEven(px*100)) < 1e-6 {
		return true
	}
	if math.Abs(px*10000-math.RoundToEven(px*10000)) < 1e-4 {
		return true
	}
	return false
}

func inferStandaloneOptionQty(cash float64) float64 {
	absCash := math.Abs(cash)
	if !(absCash > 0) {
		return 0
	}
	maxQty := py.RoundInt(absCash)
	if maxQty < 1 {
		maxQty = 1
	}
	if maxQty > 10000 {
		maxQty = 10000
	}
	for q := 1; q <= maxQty; q++ {
		if isCleanOptionQty(absCash, float64(q)) {
			return float64(q)
		}
	}
	return 1
}

func pickCleanQty(cash, openSz float64) float64 {
	picked := 0.0
	cap := int(openSz + 1e-9)
	if cap < 1 {
		cap = 1
	}
	for q := 1; q <= cap; q++ {
		if isCleanOptionQty(cash, float64(q)) {
			picked = float64(q)
			break
		}
	}
	qty := picked
	if qty == 0 {
		qty = inferStandaloneOptionQty(cash)
	}
	if qty > openSz {
		qty = openSz
	}
	return qty
}

func inferZeroQtyOptionFills(fills []*fill) {
	remaining := map[string]*dirs{}
	remOf := func(a *Act) *dirs {
		k := bookKey(a)
		r, ok := remaining[k]
		if !ok {
			r = &dirs{}
			remaining[k] = r
		}
		return r
	}
	zerosByBook := map[string][]int{}
	pools := map[rollKey]*dirs{}
	poolOf := func(a *Act) *dirs {
		k := rollKeyOf(a)
		p, ok := pools[k]
		if !ok {
			p = &dirs{}
			pools[k] = p
		}
		return p
	}
	upcoming := func(k string, i int) int {
		n := 0
		for _, j := range zerosByBook[k] {
			if j > i {
				n++
			}
		}
		return n
	}

	for i, f := range fills {
		a := f.a
		qty := math.Abs(a.Quantity)
		cash := a.NetCashAmount
		if !isOption(a.Symbol) || f.side == "" {
			continue
		}
		if qty == 0 && math.Abs(cash) > 1e-9 {
			k := bookKey(a)
			zerosByBook[k] = append(zerosByBook[k], i)
		}
	}

	for i, f := range fills {
		a := f.a
		if f.side == "" {
			continue
		}
		rem := remOf(a)
		if isOption(a.Symbol) && isMultileg(a) {
			pool := poolOf(a)
			var direction string
			switch {
			case rem.short > EPS:
				direction = "SHORT"
			case rem.long > EPS:
				direction = "LONG"
			case pool.short >= pool.long:
				direction = "SHORT"
			default:
				direction = "LONG"
			}
			openSz := *rem.at(direction) + *pool.at(direction)
			qty := math.Abs(a.Quantity)
			cash := a.NetCashAmount
			if qty == 0 {
				k := bookKey(a)
				if *rem.at(direction) > EPS && upcoming(k, i) == 0 {
					qty = *rem.at(direction)
				} else if openSz > EPS {
					qty = pickCleanQty(cash, openSz)
				} else {
					qty = inferStandaloneOptionQty(cash)
				}
				if qty > 0 {
					a.UnitPrice = math.Abs(cash) / (qty * 100.0)
				} else {
					a.UnitPrice = 0
				}
			}
			if direction == "SHORT" {
				f.side = "BUY"
				a.ActivitySubType = "BUYTOCLOSE"
				a.Quantity = qty
			} else {
				f.side = "SELL"
				a.ActivitySubType = "SELLTOCLOSE"
				a.Quantity = -qty
			}
			f.qty = qty
			f.rollDirection = direction
			closed := math.Min(qty, *rem.at(direction))
			*rem.at(direction) -= closed
			*pool.at(direction) -= math.Min(qty-closed, *pool.at(direction))
			if closed > EPS || qty > EPS {
				*pool.at(direction) += qty
			}
			continue
		}
		if isOption(a.Symbol) {
			resolveOptionFillSide(f, rem)
		}
		qty := math.Abs(a.Quantity)
		cash := a.NetCashAmount
		if isOption(a.Symbol) && qty == 0 {
			closingDir := "LONG"
			if f.side == "BUY" {
				closingDir = "SHORT"
			}
			openSz := *rem.at(closingDir)
			if math.Abs(cash) > 1e-9 {
				k := bookKey(a)
				if openSz > 0 && upcoming(k, i) == 0 {
					qty = openSz
				} else if openSz > 0 {
					qty = pickCleanQty(cash, openSz)
				} else {
					qty = inferStandaloneOptionQty(cash)
				}
				if qty > 0 {
					a.UnitPrice = math.Abs(cash) / (qty * 100.0)
				} else {
					a.UnitPrice = 0
				}
			} else if openSz > 0 && isCloseOnly(a) {
				qty = openSz
				a.UnitPrice = 0
			}
			if qty > 0 {
				if f.side == "SELL" {
					a.Quantity = -qty
				} else {
					a.Quantity = qty
				}
				f.qty = qty
			}
		}
		if isOption(a.Symbol) && (strings.Contains(compact(a.RawType), "ASSIGN") || strings.Contains(compact(a.ActivityType), "ASSIGN")) {
			a.UnitPrice = 0
		}
		if f.qty > 0 {
			closingDir := "LONG"
			if f.side == "BUY" {
				closingDir = "SHORT"
			}
			opening := openingDirection(a, f.side)
			left := f.qty
			closeAmt := math.Min(left, *rem.at(closingDir))
			*rem.at(closingDir) -= closeAmt
			left -= closeAmt
			if left > EPS && isOption(a.Symbol) {
				pool := poolOf(a)
				pooled := math.Min(left, *pool.at(closingDir))
				*pool.at(closingDir) -= pooled
				left -= pooled
			}
			if left > EPS && opening != "" {
				*rem.at(opening) += left
			}
		}
	}
}

File: crates/model/src/normalize.rs
//! Activity normalization: a copy of each raw Wealthsimple row with crypto and
//! option events expressed as trade fills. `model.normalize_activity`.
//!
//! The raw rows are never rewritten in the store; this is a derived copy.

use serde_json::{Map, Value};

use crate::symbols::{is_option_symbol, option_right, underlying_symbol};
use crate::value::{compact, field_num, field_s, norm_account_name, num, EPS};

pub const KINDS: [&str; 4] = ["Shares", "Options", "Crypto", "Futures"];

/// `model.is_crypto_activity`.
pub fn is_crypto_activity(a: &Value) -> bool {
    compact(&field_s(a, "rawType")).starts_with("CRYPTO")
        || compact(&field_s(a, "activityType")).starts_with("CRYPTO")
}

/// `model.kind_of`: an explicit kind wins, then crypto, then the symbol.
pub fn kind_of(a: &Value) -> String {
    let k = field_s(a, "kind");
    if KINDS.contains(&k.as_str()) { return k; }
    if is_crypto_activity(a) { return "Crypto".into(); }
    if is_option_symbol(&field_s(a, "symbol")) { return "Options".into(); }
    "Shares".into()
}

fn type_fields(a: &Value) -> (String, String) {
    (compact(&field_s(a, "activityType")), compact(&field_s(a, "activitySubType")))
}

/// `model.is_intentional_open`: a row that says it opens a position.
pub fn is_intentional_open(a: &Value) -> bool {
    let (at, sub) = type_fields(a);
    at.contains("TOOPEN") || sub.contains("TOOPEN") || at == "STO" || at == "BTO" || sub == "STO" || sub == "BTO"
}

/// `model.is_close_only`: a row that can only reduce a position -- an explicit
/// close, or the expiry/assignment/exercise the broker posts for one.
pub fn is_close_only(a: &Value) -> bool {
    let (at, sub) = type_fields(a);
    for f in [&at, &sub] {
        if f.contains("TOCLOSE") || f == "BTC" || f == "STC" { return true; }
    }
    for f in [&at, &sub] {
        if f.contains("EXPIR") || f.contains("ASSIGN") || f.contains("EXERCISE") { return true; }
    }
    false
}

/// `model.opening_direction`: which way a fill opens, or `None` when it can
/// only close. A bare share sale is never read as a short unless the row says
/// it opened one, so a sale of something bought before the history starts does
/// not invent a short position.
pub fn opening_direction(a: &Value, side: &str) -> Option<&'static str> {
    match side {
        "BUY" => if is_close_only(a) { None } else { Some("LONG") },
        "SELL" => {
            if is_option_symbol(&field_s(a, "symbol")) {
                return if is_close_only(a) { None } else { Some("SHORT") };
            }
            if is_intentional_open(a) { return Some("SHORT"); }
            None
        }
        _ => None,
    }
}

fn set(m: &mut Map<String, Value>, k: &str, v: Value) { m.insert(k.into(), v); }
fn setf(m: &mut Map<String, Value>, k: &str, v: f64) {
    set(m, k, Value::Number(serde_json::Number::from_f64(v).unwrap_or_else(|| 0.into())));
}
fn push_flag(m: &mut Map<String, Value>, flag: &str) {
    if let Some(Value::Array(a)) = m.get_mut("flags") { a.push(Value::String(flag.into())); }
}

/// `model.normalize_activity`.
pub fn normalize_activity(activity: &Value) -> Value {
    let mut a: Map<String, Value> = match activity {
        Value::Object(o) => o.clone(),
        _ => Map::new(),
    };
    let account = norm_account_name(&field_s(activity, "accountType"));
    set(&mut a, "accountType", Value::String(account));
    let rt = compact(&field_s(activity, "rawType"));
    let at = compact(&field_s(activity, "activityType"));
    let cash = field_num(activity, "netCashAmount");
    let qty = field_num(activity, "quantity").abs();
    set(&mut a, "flags", Value::Array(vec![]));

    // ---- crypto: the rows carry their direction in the type, not the sign
    if rt == "CRYPTOBUY" || at == "CRYPTOBUY" {
        set(&mut a, "category", "trade".into());
        set(&mut a, "activityType", "Trade".into());
        set(&mut a, "activitySubType", "BUY".into());
        set(&mut a, "kind", "Crypto".into());
        setf(&mut a, "quantity", qty);
        setf(&mut a, "netCashAmount", -cash.abs());
        return Value::Object(a);
    }
    if rt == "CRYPTOSELL" || at == "CRYPTOSELL" {
        set(&mut a, "category", "trade".into());
        set(&mut a, "activityType", "Trade".into());
        set(&mut a, "activitySubType", "SELL".into());
        set(&mut a, "kind", "Crypto".into());
        setf(&mut a, "quantity", -qty);
        setf(&mut a, "netCashAmount", cash.abs());
        return Value::Object(a);
    }
    if rt == "CRYPTOTRANSFER" || at == "CRYPTOTRANSFER" {
        let sub = compact(&field_s(activity, "activitySubType"));
        set(&mut a, "category", "trade".into());
        set(&mut a, "activityType", "Trade".into());
        set(&mut a, "kind", "Crypto".into());
        push_flag(&mut a, "transfer");
        if sub.contains("OUT") || cash < 0.0 {
            set(&mut a, "activitySubType", "SELL".into());
            push_flag(&mut a, "transfer-out");
            setf(&mut a, "quantity", -qty);
            setf(&mut a, "netCashAmount", cash.abs());
        } else {
            set(&mut a, "activitySubType", "BUY".into());
            setf(&mut a, "quantity", qty);
            setf(&mut a, "netCashAmount", -cash.abs());
        }
        return Value::Object(a);
    }
    if rt == "CRYPTOSTAKINGREWARD" || at == "CRYPTOSTAKINGREWARD" {
        // Units arriving at no cost: they enter the book at zero, so the whole
        // proceeds show as gain when they are sold.
        set(&mut a, "category", "trade".into());
        set(&mut a, "activityType", "Trade".into());
        set(&mut a, "activitySubType", "BUY".into());
        set(&mut a, "kind", "Crypto".into());
        push_flag(&mut a, "reward");
        setf(&mut a, "quantity", qty);
        setf(&mut a, "unitPrice", 0.0);
        setf(&mut a, "netCashAmount", 0.0);
        return Value::Object(a);
    }
    if rt.starts_with("CRYPTO") {
        set(&mut a, "category", "other".into());
        set(&mut a, "kind", "Crypto".into());
        return Value::Object(a);
    }

    // A distribution posted in units with no cash is a pending notice, not a
    // share delivery: Wealthsimple's balance does not grow by it.
    if at == "STKDIS" && rt == "DIVIDEND" && cash.abs() < EPS {
        set(&mut a, "category", "other".into());
        push_flag(&mut a, "pending-distribution");
        return Value::Object(a);
    }

    let raw = format!("{}{}", rt, at);
    if raw.contains("MULTILEG") {
        set(&mut a, "category", "trade".into());
        if cash < 0.0 || compact(&field_s(activity, "direction")) == "DEBIT" {
            set(&mut a, "activityType", "OPTIONS_BUY".into());
            set(&mut a, "activitySubType", "BUYTOCLOSE".into());
        } else {
            set(&mut a, "activityType", "OPTIONS_SELL".into());
            set(&mut a, "activitySubType", "SELLTOOPEN".into());
        }
    } else if raw.contains("EXPIR") || raw.contains("ASSIGN") || raw.contains("EXERCISE") {
        set(&mut a, "category", "option_event".into());
        if raw.contains("ASSIGN") {
            set(&mut a, "activityType", "ASSIGN".into());
            set(&mut a, "activitySubType", "BUYTOCLOSE".into());
            setf(&mut a, "unitPrice", 0.0);
        } else if raw.contains("SHORTEXPIR") {
            set(&mut a, "activityType", "EXPIR".into());
            set(&mut a, "activitySubType", "BUY".into());
        } else if raw.contains("EXPIR") {
            set(&mut a, "activityType", "EXPIR".into());
            set(&mut a, "activitySubType", "SELL".into());
        } else {
            set(&mut a, "activityType", "EXERCISE".into());
            set(&mut a, "activitySubType", "SELL".into());
        }
        if raw.contains("ASSIGN") || cash.abs() < 1e-12 {
            setf(&mut a, "unitPrice", 0.0);
        }
        if qty > 0.0 {
            let sub = a.get("activitySubType").map(|v| v.as_str().unwrap_or("")).unwrap_or("");
            let q = if sub == "SELL" { -qty } else { qty };
            setf(&mut a, "quantity", q);
        }
    }
    let kind = kind_of(&Value::Object(a.clone()));
    set(&mut a, "kind", Value::String(kind));
    Value::Object(a)
}

pub fn normalize_activities(activities: &[Value]) -> Vec<Value> {
    activities.iter().map(normalize_activity).collect()
}

/// `model.fifo_account`: the nickname when there is one, so two accounts with
/// the same symbol keep separate books; the ids only when there is not.
pub fn fifo_account(a: &Value) -> String {
    let nick = norm_account_name(&field_s(a, "accountType"));
    if !nick.is_empty() { return nick; }
    let fifo = field_s(a, "fifoId");
    if !fifo.is_empty() { return fifo; }
    field_s(a, "accountId")
}

/// `model.book_key`.
pub fn book_key(a: &Value) -> String {
    format!("{}::{}::{}", fifo_account(a), field_s(a, "symbol"), field_s(a, "currency"))
}

/// `model.roll_key`: what an option roll is folded within -- one account, one
/// underlying, one right.
pub fn roll_key(a: &Value) -> (String, String, &'static str) {
    let sym = field_s(a, "symbol");
    (fifo_account(a), underlying_symbol(&sym), option_right(&sym))
}

/// `model.is_multileg`.
pub fn is_multileg(a: &Value) -> bool {
    compact(&field_s(a, "rawType")).contains("MULTILEG")
}

/// `model.fold_stkdis`: net the +N/-N name-change rows posted on one day, and
/// open whatever is left over at $0.
pub fn fold_stkdis(activities: &[Value]) -> Vec<Value> {
    struct G { pos: f64, neg: f64, sample: Value }
    let mut rest: Vec<Value> = Vec::new();
    let mut groups: Vec<(String, G)> = Vec::new();

    for a in activities {
        if compact(&field_s(a, "activityType")) != "STKDIS" {
            rest.push(a.clone());
            continue;
        }
        let k = format!("{}\u{0}{}\u{0}{}", field_s(a, "symbol"), field_s(a, "transactionDate"), field_s(a, "currency"));
        let idx = match groups.iter().position(|(gk, _)| *gk == k) {
            Some(i) => i,
            None => { groups.push((k, G { pos: 0.0, neg: 0.0, sample: a.clone() })); groups.len() - 1 }
        };
        let q = field_num(a, "quantity");
        if field_s(a, "activitySubType") == "SELL" || q < 0.0 {
            groups[idx].1.neg += q.abs();
        } else {
            groups[idx].1.pos += q.abs();
        }
    }
    for (_, g) in groups {
        let net = g.pos - g.neg;
        if net > EPS {
            let mut m = match g.sample { Value::Object(o) => o, _ => Map::new() };
            setf(&mut m, "quantity", net);
            set(&mut m, "activitySubType", "BUY".into());
            setf(&mut m, "unitPrice", 0.0);
            setf(&mut m, "netCashAmount", 0.0);
            set(&mut m, "category", "trade".into());
            rest.push(Value::Object(m));
        }
    }
    rest
}

/// `model.num` re-exported for callers that hold a raw `Value`.
pub fn n(v: Option<&Value>) -> f64 { num(v, 0.0) }
